src/hdbscan.jl

```diff
@@ -0,0 +1,311 @@
+
+"""
+    isnoise(c::HdbscanCluster)
+This function returns whether the cluster is the noise or not.
+"""
+isnoise(c::HdbscanCluster) = c.stability == -1
+isstable(c::HdbscanCluster) = c.stability != 0
+function increment_stability(c::HdbscanCluster, λbirth)
+    c.stability += sum(c.λp) - length(c.λp) * λbirth
+end
+
+"""
+    HdbscanResult 
+Result of the [hdbscan](@ref) clustering.
+- `clusters`: vector of clusters
+- `assignments`: vectors of assignments for each points
+"""
+struct HdbscanResult
+    clusters::Vector{HdbscanCluster}
+    assignments::Vector{Int}
+end
+
+"""
+    hdbscan(points::AbstractMatrix, ncore::Integer, min_cluster_size::Integer;
+            metric=SqEuclidean())
+Cluster `points` using Density-Based Clustering Based on Hierarchical Density Estimates (HDBSCAN) algorithm.
+Refer to [HDBSCAN algorithm](@ref hdbscan_algorithm) description for the details on how the algorithm works.
+# Parameters
+- `points`: the *d*×*n* matrix, where each column is a *d*-dimensional point
+- `ncore::Integer`: number of *core* neighbors of point, see [HDBSCAN algorithm](@ref hdbscan_algorithm) for a description
+- `min_cluster_size::Integer`: minimum number of points in the cluster
+- `metric`(defaults to Euclidean): the points distance metric to use.
+"""
+function hdbscan(points::AbstractMatrix, k::Int, min_cluster_size::Int; metric=Euclidean())
+    if min_cluster_size < 1
+        throw(DomainError(min_cluster_size, "The `min_cluster_size` must be greater than or equal to 1"))
+    end
+    n = size(points, 2)
+    dists = pairwise(metric, points; dims=2)
+    #calculate core distances for each point
+    core_dists = core_distances(dists, k)
+    #calculate mutual reachability distance between any two points
+    mrd = hdbscan_graph(core_dists, dists)
+    #compute a minimum spanning tree by prim method
+    mst = hdbscan_minspantree(mrd, n)
+    #build a HDBSCAN hierarchy
+    hierarchy = hdbscan_clusters(mst, min_cluster_size)
+    #extract the target cluster
+    prune_cluster!(hierarchy)
+    #generate the list of cluster assignment for each point
+    result = HdbscanCluster[]
+    noise_points = fill(-1, n)
+    for (i, j) in enumerate(hierarchy[2n-1].children)
+        c = hierarchy[j]
+        push!(result, c)
+        for k in c.points
+            noise_points[k] = 0
+        end
+    end
+    push!(result, HdbscanCluster(Int[]))
+    result[end].points = findall(x->x==-1, noise_points)
+    assignments = Array{Int}(undef, length(points))
+    for i in 1 : length(result)-1
+        assignments[result[i].points] = i
+    end
+    assignments[result[end].points] .= -1
+    return HdbscanResult(result, assignments)
+end
+
+# calculate the core distances of the points
+function core_distances(dists::AbstractMatrix, k::Integer)
+    core_dists = Array{Float64}(undef, size(dists, 1))
+    for i in axes(dists, 1)
+        dist = sort(dists[i, :])
+        core_dists[i] = dist[k]
+    end
```

Comment:
It just occurred to me that it could be optimized into
```suggestion
    for i in axes(dists, 2)
        core_dists[i] = partialsort!(dists[:, i], k)
    end
```

Note that in Julia iterating over columns is much more efficient than by rows, because one column occupies a continuous block of memory, whereas in the row the neighboring elements are stored *N* elements (the number of columns) apart.